File: common/platform-dotnet/SoundMetrics.Data/FineDuration.fs
// Copyright 2011-2019 Sound Metrics Corp. All Rights Reserved.

namespace SoundMetrics.Data

open System

[<Struct>]
type FineDuration private (microseconds : float) =

    static member Zero = FineDuration 0.0
    static member OneMicrosecond = FineDuration 1.0

    static member FromMicroseconds (microseconds : float) = FineDuration microseconds
    static member FromMilliseconds (milliseconds : float) = FineDuration (milliseconds * 1000.0)

    member __.TotalMicroseconds = microseconds
    member __.TotalMilliseconds = microseconds / 1000.0
    member __.TotalSeconds = microseconds / 1000000.0

    member __.Floor = Math.Floor(microseconds)
    member __.Ceiling = Math.Ceiling(microseconds)

    override this.ToString() = String.Format("{0:0.000000} s", this.TotalSeconds)

    static member (/) (d : FineDuration, divisor) : FineDuration =
        FineDuration (d.TotalMicroseconds / divisor)
    static member (*) (d : FineDuration, multiplier) : FineDuration =
        FineDuration (d.TotalMicroseconds * multiplier)
    static member (*) (multiplier, d : FineDuration) : FineDuration =
        FineDuration (d.TotalMicroseconds * multiplier)

    static member (+) (d1 : FineDuration, d2 : FineDuration) : FineDuration =
        FineDuration (d1.TotalMicroseconds + d2.TotalMicroseconds)
    static member (-) (d : FineDuration, subtrahend : FineDuration) : FineDuration =
        FineDuration (d.TotalMicroseconds - subtrahend.TotalMicroseconds)

    // op_LessThan for interop with other CLI languages.
    static member op_LessThan(a : FineDuration, b : FineDuration) = a < b
    // op_GreaterThan for interop with other CLI languages.
    static member op_GreaterThan(a : FineDuration, b : FineDuration) = a > b
    // op_Equality for interop with other CLI languages.
    static member op_Equality(a : FineDuration, b : FineDuration) = (a = b)
    // op_Inequality for interop with other CLI languages.
    static member op_Inequality(a : FineDuration, b : FineDuration) = (a <> b)
    // op_GreaterThanOrEqual for interop with other CLI languages.
    static member op_GreaterThanOrEqual(a : FineDuration, b : FineDuration) = a >= b
    // op_LessThanOrEqual for interop with other CLI languages.
    static member op_LessThanOrEqual(a : FineDuration, b : FineDuration) = a <= b

    /// Returns the minimum of two values. C# implements Math.Min/Max as overloads
    /// rather than a comparable generic (dates to .NET 1.1, so before generics).
    /// F# implements min/max as comparable generics.
    /// (Extension methods have instance-method semantics, so aren't viable for
    /// extending Math.Min/Max.
    static member Min(a : FineDuration, b : FineDuration) =
        if a < b then a else b

    /// Returns the maximum of two values. C# implements Math.Min/Max as overloads
    /// rather than a comparable generic (dates to .NET 1.1, so before generics).
    /// F# implements min/max as comparable generics.
    /// (Extension methods have instance-method semantics, so aren't viable for
    /// extending Math.Min/Max.
    static member Max(a : FineDuration, b : FineDuration) =
        if a > b then a else b

File: common/platform-dotnet/SoundMetrics.Data2/FineDuration.cs
// Copyright 2011-2020 Sound Metrics Corp. All Rights Reserved.

using System;

namespace SoundMetrics.Data
{
    public struct FineDuration
        : IComparable, IComparable<FineDuration>, IEquatable<FineDuration>
    {
        public static FineDuration Zero = FromMicroseconds(0.0);
        public static FineDuration OneMicrosecond = FromMicroseconds(1.0);

        public static FineDuration FromMicroseconds(double microseconds) =>
            new FineDuration(microseconds);

        public static FineDuration FromMilliseconds(double milliseconds) =>
            new FineDuration(milliseconds * 1_000.0);

        public double TotalMicroseconds => microseconds;
        public double TotalMilliseconds => microseconds / 1_000.0;
        public double TotalSeconds => microseconds / 1_000_000.0;

        public static FineDuration operator +(FineDuration d) => d;
        public static FineDuration operator -(FineDuration d) => FromMicroseconds(-d.microseconds);

        public static FineDuration operator +(FineDuration d1, FineDuration d2) =>
            FromMicroseconds(d1.TotalMicroseconds + d2.TotalMicroseconds);
        public static FineDuration operator -(FineDuration d1, FineDuration d2) =>
            FromMicroseconds(d1.TotalMicroseconds - d2.TotalMicroseconds);

        public static FineDuration operator *(FineDuration d, double multiplier)
            => FromMicroseconds(d.TotalMicroseconds * multiplier);
        public static FineDuration operator *(double multiplier, FineDuration d)
            => FromMicroseconds(d.TotalMicroseconds * multiplier);

        public static FineDuration operator /(FineDuration d, double divisor)
        {
            if (divisor == 0.0)
            {
                throw new DivideByZeroException();
            }

            return FromMicroseconds(d.TotalMicroseconds / divisor);
        }

        public static FineDuration operator /(double f, FineDuration divisor)
        {
            if (divisor == Zero)
            {
                throw new DivideByZeroException();
            }

            return FromMicroseconds(f / divisor.TotalMicroseconds);
        }

        public static bool operator ==(FineDuration a, FineDuration b) =>
            a.TotalMicroseconds == b.TotalMicroseconds;
        public static bool operator !=(FineDuration a, FineDuration b) =>
            a.TotalMicroseconds != b.TotalMicroseconds;

        public static bool operator <(FineDuration a, FineDuration b) =>
            a.microseconds < b.microseconds;
        public static bool operator <=(FineDuration a, FineDuration b) =>
            a.microseconds <= b.microseconds;
        public static bool operator >=(FineDuration a, FineDuration b) =>
            a.microseconds >= b.microseconds;
        public static bool operator >(FineDuration a, FineDuration b) =>
            a.microseconds > b.microseconds;

        public override string ToString() =>
            $"{microseconds:0.000} \u00b5s";

        public bool Equals(FineDuration other)
            => this.microseconds == other.microseconds;

        public int CompareTo(object obj)
        {
            if (obj is FineDuration other)
            {
                return ((IComparable<double>)microseconds).CompareTo(other.microseconds);
            }
            else
            {
                return obj.Equals(this) ? 0 : 1;
            }
        }

        public int CompareTo(FineDuration other) =>
            ((IComparable<double>)microseconds).CompareTo(other.microseconds);

        public override bool Equals(object obj)
        {
            return obj is FineDuration duration &&
                   microseconds == duration.microseconds;
        }

        public override int GetHashCode()
        {
            return microseconds.GetHashCode();
        }

        private FineDuration(double microseconds)
        {
            this.microseconds = microseconds;
        }

        private readonly double microseconds;
    }
}
